lib/puppetserver/ca/generate_action.rb

```diff
@@ -0,0 +1,249 @@
+        Puppetserver::Utils::FileUtilities.ensure_file(puppet.settings[:cert_inventory], "", 0640)
+
+        @logger.inform "Generation succeeded. Find your files in #{puppet.settings[:cadir]}"
+        return 0
+      end
+
+      def log_errors_and_return(errors)
+        errors = Array(errors).compact
+        @logger.err "Error:"
+        errors.each do |message|
+          @logger.err "    #{message}"
+        end
+        return 1
+      end
+
+      def generate_root_and_intermediate_ca(puppet)
+        private_key_length = puppet.settings[:keylength]
+        root_ca_name = puppet.settings[:root_ca_name]
+        int_ca_name = puppet.settings[:ca_name]
+
+        ttl_setting = Puppetserver::Settings::TTLSetting.new(:ca_ttl, puppet.settings[:ca_ttl])
+        log_errors_and_return(ttl_setting.errors) if ttl_setting.errors.any?
+        ca_ttl = ttl_setting.munged_value
+
+        root_key = create_private_key(private_key_length)
+        root_cert = self_signed_ca(root_key, root_ca_name, ca_ttl)
+        root_crl = create_crl_for(root_cert, root_key, ca_ttl)
+
+        int_key = create_private_key(private_key_length)
+        int_csr = create_csr(int_key, int_ca_name)
+        int_cert = sign(root_key, root_cert, int_csr, ca_ttl)
+        int_crl = create_crl_for(int_cert, int_key, ca_ttl)
+
+        Puppetserver::Utils::FileUtilities.ensure_dir(puppet.settings[:cadir])
+
+        future_files = [ { location: puppet.settings[:cacert], content: [int_cert, root_cert]},
+                         { location: puppet.settings[:cakey], content: int_key },
+                         { location: puppet.settings[:rootkey], content: root_key },
+                         { location: puppet.settings[:cacrl], content: [int_crl, root_crl]}
+                       ]
+
+        future_files.each do | file|
+          @logger.warn "#{file[:location]} exists, overwriting" if File.exist?(file[:location])
+          Puppetserver::Utils::FileUtilities.write_file(file[:location], file[:content], 0640)
+        end
+      end
+
+      def create_private_key(length)
+        OpenSSL::PKey::RSA.new(length.to_i)
+      end
+
+      def self_signed_ca(key, name, ca_ttl)
+        cert = OpenSSL::X509::Certificate.new
+
+        cert.public_key = key.public_key
+        cert.subject = OpenSSL::X509::Name.new([["CN", name]])
+        cert.issuer = cert.subject
+        cert.version = 2
+        cert.serial = 1
+
+        # Make the certificate valid as of yesterday, because so many people's
+        # clocks are out of sync.  This gives one more day of validity than people
+        # might expect, but is better than making every person who has a messed up
+        # clock fail, and better than having every cert we generate expire a day
+        # before the user expected it to when they asked for "one year".
+        cert.not_before = Time.now - (60*60*24)
+        cert.not_after  = Time.now + ca_ttl
```

Comment:
We do this when creating the root cert but the CRL is set to one second in the past and the intermediate cert is just set to `Time.now`. Can we save these values and then use them everywhere? Do the Int. cert & CRLs need to have a not_before time later than the issuing cert?

Reply:
I can set the ca_cert and int_cert to both use "yesterday".  The int_cert is always created after the ca_cert, so it will have a not_before of later than the issuing cert.  With the crl, I don't know.  I know we only do -1 second here: https://github.com/puppetlabs/puppet/blob/09bdfa9eaa368c241e68f01fb270b03aab6fe7b4/lib/puppet/ssl/certificate_revocation_list.rb#L88

Reply:
^^ talked to Justin and We'll use the same value for crl and certs.  Checked this and it doesn't seem to cause any problems with them all having the same last_updated time.